research/json-crdt.md

```diff
@@ -0,0 +1,289 @@
+
+A register as defined in the paper is represented as a multi-value CRDT register
+which keeps a mapping between the register values and the operation ID that set
+it.
+
+## Diving deep 
+
+In this section we'll describe in more detail the JSON CRDT data structure and
+algorithm that ensure strong eventual consistency and no user input loss across 
+all replicas.
+
+### Editing the JSON document and replica local state
+
+Each replica keeps its own state of the document. Changes in the document 
+occur only locally. When the document is edited, it generates an `operation` 
+which uniquely describes the mutation globally. The operations are distributed 
+to all peers as a mechanism to reach state consistency across all peers.
+
+Once a replica receives an external operation, it applies the operation locally. 
+The merging algorithm ensures that applying external operations locally will 
+not conflict with the current local state.
+
+The state of the document at any point in time is a set of operations which were
+applied locally.
+
+### Operations
+
+The paper defines the JSON CRDT as an operation-based CRDT, in which an 
+operation - a tuple representing uniquely a mutation in the JSON document - can be
+propagated to other replicas without sending the whole local document state. The
+local state of the document is represented by a set of individual operations.
+
+The operation-based representation has the advantage that replicas need only to
+share and merge singular operations, instead of the entire local state. On the 
+other hand, whenever a new node joins the network, it still needs to receive the
+whole document state, which grows linearly with the number of operations applied
+to the document. This problem can be addressed with, for example,
+[Delta State CRDTs](https://github.com/ipfs/research-CRDT/issues/31).
+
+#### Operation Representation
+
+An operation is a tuple of the form
+
+```
+op (
+  id: N x ReplicaID
+  deps: P(N x ReplicaID)
+  cur: cursor(<k1,...,kn-1>, kn)
+  mut: insert(v) | delete | assign(v)   v: VAL
+)
+```
+
+where the `id` is a Lamport timestamp which identifies the operation uniquely; 
+`deps` is a set of casual dependencies of the operation; `cur` describes the 
+position of the document to be modified; and `mut` describes the modification 
+itself.
+
+- **Id**
+
+The operation `id` is a Lamport timestamp (see appendix), which guarantees that
+every operation is uniquely identified globally and that the peers implicitly
+agree on the partial casual order of when the operations occurred, without any 
+other synchronization mechanism.
+
+- **Cursor**
+
+A cursor `cur` describes the path from the root to a branch or leaf of the 
+document in which the mutation will be applied. The cursor must consist only of 
+immutable keys and identifiers.
+
+If we consider the following JSON document
+
+```
+{"shopping": ["cheese", "eggs", "milk"]}
+```
+
+the cursor
+
+```
+cursor(<MapT(doc), listT("shopping")>, id1)
```

Comment:
What is the value of `id1` here?

Reply:
- [ ] replace `id1` to something that makes sense given the context